src/Features/Core/Portable/UseCompoundAssignment/AbstractUseCompoundAssignmentDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+
+namespace Microsoft.CodeAnalysis.UseCompoundAssignment
+{
+    internal abstract class AbstractUseCompoundAssignmentDiagnosticAnalyzer<
+        TSyntaxKind,
+        TAssignmentSyntax,
+        TBinaryExpressionSyntax>
+        : AbstractCodeStyleDiagnosticAnalyzer
+        where TSyntaxKind : struct
+        where TAssignmentSyntax : SyntaxNode
+        where TBinaryExpressionSyntax : SyntaxNode
+    {
+        private readonly ISyntaxFactsService _syntaxFacts;
+        protected readonly ImmutableDictionary<TSyntaxKind, TSyntaxKind> BinaryToAssignmentMap;
+        protected readonly ImmutableDictionary<TSyntaxKind, TSyntaxKind> AssignmentToTokenMap;
+
+        protected AbstractUseCompoundAssignmentDiagnosticAnalyzer(
+            ISyntaxFactsService syntaxFacts,
+            ImmutableDictionary<TSyntaxKind, TSyntaxKind> binaryToAssignmentMap,
+            ImmutableDictionary<TSyntaxKind, TSyntaxKind> assignmentToTokenMap)
+            : base(IDEDiagnosticIds.UseCompoundAssignmentDiagnosticId,
+                   new LocalizableResourceString(
+                       nameof(FeaturesResources.Use_compound_assignment), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+            _syntaxFacts = syntaxFacts;
+            BinaryToAssignmentMap = binaryToAssignmentMap;
+            AssignmentToTokenMap = assignmentToTokenMap;
+
+            Debug.Assert(BinaryToAssignmentMap.Count == AssignmentToTokenMap.Count);
+            Debug.Assert(BinaryToAssignmentMap.Values.All(AssignmentToTokenMap.ContainsKey));
+        }
+
+        protected abstract TSyntaxKind GetKind(int rawKind);
+        protected abstract TSyntaxKind GetAnalysisKind();
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeAssignment, GetAnalysisKind());
+
+        private void AnalyzeAssignment(SyntaxNodeAnalysisContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var assignment = (TAssignmentSyntax)context.Node;
+
+            var options = context.Options;
+            var syntaxTree = assignment.SyntaxTree;
+            var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
```

Comment:
>.GetAwaiter().GetResult() [](start = 92, length = 25)

This surprised me, as I didn't encounter it in other analyzers. Don't we have a version of `RegisterSyntaxNodeAction` that would allow for `AnalyzeAssignment` to be made async, so this becomes `await`?

Reply:
> that would allow for AnalyzeAssignment to be made async, so this becomes await

No.  All analysis is synchronous. That's baked *heavily* into the compiler's analyzer pipeline.  So this is hte workaround (for now).  Sam/Heejae/Manish are working on a better solution.